### ISSUE TITLE
AuthenticationFailed error when accessing table using SAS

### ISSUE THREAD

**Issue by enkidu123** [Azure/azure-storage-python#137]
I keep getting Authentication errors when trying to access a table with a generated SAS - as far as I can tell I'm following the same approach as the table_sas() example in the samples folder. (sensitive info replaced with 'XXX'):
```python
from azure.storage.table import TableService, TablePermissions
from datetime import datetime, timedelta

ACCOUNT = 'XXX'
TABLE = 'TestBlah'
ACCOUNT_KEY = 'XXX'

service_by_key = TableService(account_name=ACCOUNT, account_key=ACCOUNT_KEY)

service_by_key.create_table(TABLE)

# including start parameter to avoid any problems with clock skew
token = service_by_key.generate_table_shared_access_signature(
    table_name=TABLE,
    permission=TablePermissions.QUERY,
    expiry=datetime.utcnow() + timedelta(hours=10),
    start=datetime.utcnow() - timedelta(hours=10),
)

print "token:", token

service_by_sas = TableService(account_name=ACCOUNT, sas_token=token)

entities = service_by_sas.query_entities(TABLE)
for entity in entities:
    print entity
```
The authentication by key succeeds, and the table is created and SAS generated, but the subsequent SAS authentication fails:
```
azure.common.AzureHttpError: Server failed to authenticate the request. Make sure the value of Authorization header is formed correctly including the signature.
{"odata.error":{"code":"AuthenticationFailed","message":{"lang":"en-US","value":"Server failed to authenticate the request. Make sure the value of Authorization header is formed correctly including the signature.\nRequestId:dacad6f1-0002-0010-1a2c-775d73000000\nTime:2016-03-05T22:18:31.3364149Z"}}}
```
The generated SAS token looks like:
```
sp=r&sv=2015-04-05&st=2016-03-05T12%3A18%3A32Z&tn=TestBlah&sig=XXX&se=2016-03-06T08%3A18%3A32Z
```
Environment details:
Windows 7 64-bit
Python 2.7.11
azure-common 1.1.0
azure-nspkg 1.0.0
azure-storage 0.30.0

**Comment by emgerner-msft** [comment]
At least at first glance that code does look correct to me. In addition to the samples, we have [a test for this case](https://github.com/Azure/azure-storage-python/blob/master/tests/test_table_entity.py#L1018) that I've verified runs with Python 2.7 so I'm a little stumped.

Since the code looks correct, the easiest next step for us is to look into this internally. For that I need your account name, and a more recent failure (exactly like the AzureHttpError you posted above, just a more recent time). If you feel this is too sensitive to post here, please email ascl@microsoft.com with that information and I'll follow up that way.

PS: If there's no start time given, the server will only check the time is before the end time. So, not including start will prevent clock skew best. The main purpose for start time is if you actually want to enforce the SAS doesn't start immediately. What you have shouldn't be harmful though.

**Comment by enkidu123** [comment]
Ok - account is 'pabstorage', table is 'TestBlah4'. I've updated the test script to also insert an entity and query it back using the key-authenticated service.

I've tried running on a different machine/network - an Azure Linux VM - to see if there was any problem with my environment.

```python
from azure.storage.table import TableService, TablePermissions
from datetime import datetime, timedelta

ACCOUNT = 'pabstorage'
TABLE = 'TestBlah4'
ACCOUNT_KEY = 'XXX'

service_by_key = TableService(account_name=ACCOUNT, account_key=ACCOUNT_KEY)

print "*** creating table"
service_by_key.create_table(TABLE)

print "*** inserting entity"
entity = {
    'PartitionKey': 'John',
    'RowKey': 'Doe',
    'deceased': False,
}
service_by_key.insert_entity(TABLE, entity)

print "*** querying using key service"
entities = service_by_key.query_entities(TABLE)
for entity in entities:
    print entity
print 'count = ', len(list(entities))

print "*** generating SAS token"
token = service_by_key.generate_table_shared_access_signature(
    table_name=TABLE,
    permission=TablePermissions.QUERY,
    expiry=datetime.utcnow() + timedelta(minutes=5),
)

print "token:", token

print "*** querying using SAS service"
service_by_sas = TableService(account_name=ACCOUNT, sas_token=token)

entities = service_by_sas.query_entities(TABLE)
for entity in entities:
    print entity
print 'count = ', len(list(entities))
```
Output:
```
$ python test.py
*** creating table
*** inserting entity
*** querying using key service
{'Timestamp': datetime.datetime(2016, 3, 7, 20, 28, 51, 426670, tzinfo=tzlocal()), 'PartitionKey': u'John', 'etag': 'W/"datetime\'2016-03-07T20%3A28%3A51.4266708Z\'"', u'deceased': False, 'RowKey': u'Doe'}
count =  1
*** generating SAS token
token: tn=TestBlah4&sp=r&sig=kOjLPWfBp8YY9APvl7TX1Ug4S3COUzT73a3sk0nkE/A%3D&sv=2015-04-05&se=2016-03-07T20%3A33%3A51Z
*** querying using SAS service
Traceback (most recent call last):
  File "test.py", line 39, in <module>
    entities = service_by_sas.query_entities(TABLE)
  File "/home/azureuser/table-test/env/local/lib/python2.7/site-packages/azure/storage/table/tableservice.py", line 645, in query_entities
    resp = self._query_entities(*args, **kwargs)
  File "/home/azureuser/table-test/env/local/lib/python2.7/site-packages/azure/storage/table/tableservice.py", line 709, in _query_entities
    response = self._perform_request(request)
  File "/home/azureuser/table-test/env/local/lib/python2.7/site-packages/azure/storage/storageclient.py", line 125, in _perform_request
    _storage_error_handler(ex)
  File "/home/azureuser/table-test/env/local/lib/python2.7/site-packages/azure/storage/_serialization.py", line 130, in _storage_error_handler
    return _general_error_handler(http_error)
  File "/home/azureuser/table-test/env/local/lib/python2.7/site-packages/azure/storage/_error.py", line 71, in _general_error_handler
    raise AzureHttpError(message, http_error.status)
azure.common.AzureHttpError: Server failed to authenticate the request. Make sure the value of Authorization header is formed correctly including the signature.
{"odata.error":{"code":"AuthenticationFailed","message":{"lang":"en-US","value":"Server failed to authenticate the request. Make sure the value of Authorization header is formed correctly including the signature.\nRequestId:92cf7611-0002-0044-4daf-78b7f9000000\nTime:2016-03-07T20:28:51.1744913Z"}}}
```
Version information:
```
$ python --version
Python 2.7.6
$ pip freeze
argparse==1.2.1
azure-common==1.1.0
azure-nspkg==1.0.0
azure-storage==0.30.0
certifi==2016.2.28
chardet==2.3.0
enum34==1.1.2
futures==3.0.5
isodate==0.5.4
keyring==8.4.1
msrest==0.1.1
msrestazure==0.1.1
oauthlib==1.0.3
python-dateutil==2.5.0
requests==2.5.3
requests-oauthlib==0.6.1
six==1.10.0
wsgiref==0.1.2
```

**Comment by emgerner-msft** [comment]
Think I figured it out, and why I couldn't repro initially. I used a table I already had already with a *lowercase* name. It looks like when creating the sas signature we should be lowercasing the name before signing in the library, but we're not. This is an interesting edge case for the table service only since it's the only storage service that allows upper case letters (it's case preserving, but case insensitive). 

In any case, the workaround should be very simple for you. Lowercase your table name when creating the SAS:

```python
token = service_by_key.generate_table_shared_access_signature(
    table_name=TABLE.lower(),
    permission=TablePermissions.QUERY,
    expiry=datetime.utcnow() + timedelta(minutes=5),
)
```

Every other method should work fine with your existing code. We'll fix the bug in the next library version. Give it a try and tell us if that fixes the issue!

**Comment by enkidu123** [comment]
Perfect - yes that worked! Many thanks. I'll stick to lower case table names :)

**Comment by emgerner-msft** [comment]
Great, glad that worked! Thanks for the bug report and working through this with us. I'll leave this issue open and update/close it when we have the bug fix out.

**Comment by emgerner-msft** [comment]
Fixed as of 0.31.0.
